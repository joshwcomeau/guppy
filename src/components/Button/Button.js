// @flow
import React, { Component } from 'react';
import styled from 'styled-components';

import { COLORS } from '../../constants';

import CircularOutline from '../CircularOutline';

type Size = 'small' | 'medium' | 'large';
type Type = 'fill' | 'stroke';

type Props = {
  type: Type,
  size: Size,
  color1: string,
  color2: string,
  textColor?: string,
  showOutline: boolean,
  noPadding: boolean,
  style: { [key: string]: any },
  disabled: boolean,
  children: React$Node,
};

type State = {
  isHovered: boolean,
};

class Button extends Component<Props, State> {
  static defaultProps = {
    type: 'stroke',
    size: 'medium',
    color1: COLORS.purple[500],
    color2: COLORS.violet[500],
    showOutline: true,
    style: {},
  };

  getButtonElem = (size: Size) => {
    switch (size) {
      case 'small':
        return SmallButton;
      default:
      case 'medium':
        return MediumButton;
      case 'large':
        return LargeButton;
    }
  };

  render() {
    const {
      type,
      size,
      children,
      color1,
      color2,
      textColor,
      showOutline,
      style,
      disabled,
      ...delegated
    } = this.props;

    const Elem = this.getButtonElem(size);

    let mutatedStyle = { ...style };
    if (type === 'fill') {
      mutatedStyle.color = '#FFF';
      mutatedStyle.backgroundImage = `
        linear-gradient(
          45deg,
          ${color1},
          ${color2}
        )
      `;
    }

    if (textColor) {
      mutatedStyle.color = textColor;
    }

    return (
      <Elem disabled={disabled} type={type} style={mutatedStyle} {...delegated}>
        <CircularOutline
          color1={disabled ? COLORS.gray[400] : color1}
          color2={disabled ? COLORS.gray[300] : color2}
          isShown={showOutline}
          animateChanges={type === 'stroke'}
        />

        <span style={{ display: 'block' }}>{children}</span>
      </Elem>
    );
  }
}

const ButtonBase = styled.button`
  position: relative;
  border: 0;
  background: transparent;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  outline: none;
  color: ${COLORS.gray[900]};
  white-space: nowrap;

  &:not(:disabled):active rect {
    stroke-width: 4;
  }

  &:disabled {
    background-image: ${props =>
      props.type === 'fill' &&
      `linear-gradient(
      45deg,
      ${COLORS.gray[400]},
      ${COLORS.gray[300]}
    ) !important`};
  }
`;

const SmallButton = styled(ButtonBase)`
  padding: ${props => (props.noPadding ? '0px' : '0px 14px')};
  height: ${props => (props.noPadding ? 'auto' : '30px')};
  border-radius: 17px;
  font-size: 14px;
`;

const MediumButton = styled(ButtonBase)`
  padding: ${props => (props.noPadding ? '0px' : '0px 20px')};
  height: ${props => (props.noPadding ? 'auto' : '38px')};
  border-radius: 19px;
  font-size: 16px;
`;

const LargeButton = styled(ButtonBase)`
  padding: ${props => (props.noPadding ? '0px' : '0 32px')};
  height: ${props => (props.noPadding ? 'auto' : '48px')};
  border-radius: 24px;
  font-size: 24px;
`;

export default Button;
