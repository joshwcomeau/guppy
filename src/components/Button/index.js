// @flow
export { default as StrokeButton } from './StrokeButton';
export { default as FillButton } from './FillButton';

export type { Props as StrokeButtonProps } from './StrokeButton';
