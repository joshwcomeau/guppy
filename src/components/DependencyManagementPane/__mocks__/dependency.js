// @flow
export const mockReactHit = {
  name: 'React',
  version: '16.0.0',
  license: 'MIT',
  modified: 1549828927372,
  humanDownloadsLast30Days: 1000000,
};
