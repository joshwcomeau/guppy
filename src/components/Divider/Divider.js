// @flow
// Used in AddDependencySearchResult & StarterSelection
import styled from 'styled-components';

import { RAW_COLORS } from '../../constants';

const Divider = styled.div`
  width: 100%;
  height: 1px;
  background: ${RAW_COLORS.gray[100]};
`;

export default Divider;
