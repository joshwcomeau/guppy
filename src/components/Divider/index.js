// @flow
export { default } from './Divider';
