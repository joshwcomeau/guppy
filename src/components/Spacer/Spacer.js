// @flow
import styled from 'styled-components';

export default styled.div`
  display: ${props => (props.inline ? 'inline-block' : 'block')};
  width: ${props => props.size}px;
  height: ${props => props.size}px;
`;
