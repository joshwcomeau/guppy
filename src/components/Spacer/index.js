// @flow
export { default } from './Spacer';
