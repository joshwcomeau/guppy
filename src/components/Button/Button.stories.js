// @flow
import React from 'react';
import styled from 'styled-components';
import { storiesOf } from '@storybook/react';
import { decorateAction } from '@storybook/addon-actions';
import { withInfo } from '@storybook/addon-info';

import Showcase from '../../../.storybook/components/Showcase';
import Button from './Button';

const targetAction = decorateAction([args => [args[0].target]]);

const SIZES = ['xsmall', 'small', 'medium', 'large'];

storiesOf('Button', module)
  .add(
    'default',
    withInfo()(() => (
      <Button onClick={targetAction('clicked')}>Hello Button</Button>
    ))
  )
  .add(
    'sizes',
    withInfo()(() =>
      SIZES.map((size, i) => (
        <Showcase label={size} key={i}>
          <Button onClick={targetAction('button-clicked')} size={size}>
            Button
          </Button>
        </Showcase>
      ))
    )
  )
  .add(
    'types (visual styles)',
    withInfo()(() => (
      <React.Fragment>
        <Showcase label="Default (stroke)">
          <Button onClick={targetAction('button-clicked')}>Button</Button>
        </Showcase>
        <Showcase label="Fill">
          <Button onClick={targetAction('button-clicked')} type="fill">
            Button
          </Button>
        </Showcase>
      </React.Fragment>
    ))
  );
