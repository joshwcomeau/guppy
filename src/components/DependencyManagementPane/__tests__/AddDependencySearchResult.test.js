import React from 'react';
import { shallow } from 'enzyme';
import Icon from 'react-icons-kit';
import { u1F4C8 as barGraphIcon } from 'react-icons-kit/noto_emoji_regular/u1F4C8';
import lolex from 'lolex';

import {
  StatsItem,
  getColorForDownloadNumber,
  AddDependencySearchResult,
  Header,
  Description,
  StatsRow,
} from '../AddDependencySearchResult';
import Divider from '../../Divider';
import Spacer from '../../Spacer';
import { StrokeButton } from '../../Button';
import { mockReactHit } from '../__mocks__/dependency';

describe('AddDependencySearchResult component', () => {
  lolex.install();
  let wrapper;
  const projectId = 'a-project';
  const mockAddDependency = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <AddDependencySearchResult
        projectId={projectId}
        hit={mockReactHit}
        addDependency={mockAddDependency}
      />
    );
  });

  it('should render a StatsItem', () => {
    const StatsItemWrapper = shallow(
      <StatsItem icon={<Icon icon={barGraphIcon} />}>
        10 downloads a month
      </StatsItem>
    );
    expect(StatsItemWrapper).toMatchSnapshot();
  });

  // Todo: If we ever refactor the color method it would be good to create a limits constant.
  //       So we can use it here e.g. like limit[0] - 1
  //       Also exporting the color per limit would be good,
  //       so it's clear what color we expect for each limit
  // Cases to test:
  // number < 5000
  // number between 5000 & 50000
  // above 50000
  const limits = [4999, 49999, 50000];
  limits.forEach(limit => {
    it(`should create a colored download number ${limit}`, () => {
      expect(getColorForDownloadNumber(limit)).toMatchSnapshot();
    });
  });

  it('should add dependency on click', () => {
    const button = wrapper.find(Header).find(StrokeButton);
    button.simulate('click');
    expect(mockAddDependency).toHaveBeenCalledWith(
      projectId,
      mockReactHit.name,
      mockReactHit.version
    );
  });

  describe('Rendering', () => {
    it('should render a header (default)', () => {
      expect(wrapper.find(Header)).toMatchSnapshot();
    });

    it('should render a header (currentStatus idle)', () => {
      wrapper.setProps({
        currentStatus: 'idle',
      });
      expect(wrapper.find(Header)).toMatchSnapshot();
    });

    it('should render a header (currentStatus installing)', () => {
      wrapper.setProps({
        currentStatus: 'installing',
      });
      expect(wrapper.find(Header)).toMatchSnapshot();
    });

    it('should render a Description', () => {
      expect(wrapper.find(Description)).toMatchSnapshot();
    });

    it('should render a StatsRow', () => {
      expect(wrapper.find(StatsRow)).toMatchSnapshot();
    });

    it('should contain three spacers', () => {
      expect(wrapper.find(Spacer)).toHaveLength(3);
    });

    it('should contain one divider', () => {
      expect(wrapper.find(Divider)).toHaveLength(1);
    });
  });
});
