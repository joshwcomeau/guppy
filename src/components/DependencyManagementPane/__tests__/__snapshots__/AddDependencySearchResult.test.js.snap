// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddDependencySearchResult component Rendering should render a Description 1`] = `
<styled.div>
  <AlgoliaHighlighter(CustomHighlight)
    attribute="description"
    hit={
      Object {
        "humanDownloadsLast30Days": 1000000,
        "license": "MIT",
        "modified": 1549828927372,
        "name": "React",
        "version": "16.0.0",
      }
    }
  />
</styled.div>
`;

exports[`AddDependencySearchResult component Rendering should render a StatsRow 1`] = `
<styled.div>
  <StatsItem
    icon={
      Object {
        "children": Array [
          Object {
            "attribs": Object {
              "d": "M2367 2413q21 0 35.5 14.5t14.5 34.5q0 21-14.5 35.5T2367 2512H232q-21 0-35.5-14.5T182 2462V342q0-21 14.5-35.5T232 292q20 0 34.5 14.5T281 342v1644l236-241q19-21 47-21 26 0 46 16l211 177 435-1318q7-18 23-32t36-14q24 0 39.5 9t26.5 26l310 561 179-518q9-21 26-35l392-272q19-13 40-13 26 0 47 19.5t21 49.5q0 17-7 31t-22 25l-374 262-224 641q-6 19-22.5 33t-36.5 14q-28 0-42.5-9.5T1644 1350l-307-558-416 1265q-8 20-25.5 34t-40.5 14q-25 0-44-17l-241-201-289 295v231h2086z",
            },
            "children": Array [],
            "name": "path",
          },
        ],
        "viewBox": "0 0 2600 2760.837",
      }
    }
  >
    <styled.span
      style={
        Object {
          "color": "#00C853",
        }
      }
    >
      1000000
    </styled.span>
     
    downloads a month
  </StatsItem>
  <Middot />
  <License
    license="MIT"
    withIcon={true}
  />
  <Middot />
  <StatsItem
    icon={
      Object {
        "children": Array [
          Object {
            "attribs": Object {
              "d": "M2255.5 1982q-148.5 256-405 404.5T1299 2535t-551-148.5T343.5 1982 195 1431t148.5-551.5 404.5-405T1299 326t551.5 148.5 405 405T2404 1431t-148.5 551zm-92-1050.5q-134.5-230.5-365-365T1299 432q-268 0-498.5 134t-365 365T301 1431q0 267 134 497.5T799.5 2294t499.5 135 499.5-134 365-364.5T2298 1431t-134.5-499.5zM1467 1437l370 212-79 138-370-213q-19 12-41.5 19t-47.5 7q-69 0-119-50t-50-119q0-48 24.5-87.5t64.5-61.5V520h159v762q41 22 66 61.5t25 87.5q0 3-2 6zm219.5-800q19.5-18 45.5-18 28 0 45.5 19.5T1795 683q0 28-20 45.5t-45 17.5q-26 0-44.5-18t-18.5-45q0-28 19.5-46zm316.5 408.5q-19-17.5-19-45.5 0-25 17.5-45t45.5-20q23 0 43 17t20 46q0 27-18.5 46t-44.5 19q-25 0-44-17.5zm-1091-409q20 17.5 20 46.5 0 24-18 43.5T867 746q-23 0-43-17.5T804 683t18.5-46 43.5-18q26 0 46 17.5zm1178 1273q-20 16.5-43 16.5-28 0-45.5-20t-17.5-43q0-29 18.5-47t44.5-18q27 0 45 19t18 47q0 29-20 45.5zM1344 2339q-19 18-45 18t-44.5-18-18.5-45q0-26 18-45t45-19q26 0 45 19t19 45q0 27-19 45zm-863-863.5q-19 18.5-45 18.5-27 0-45-18.5t-18-44.5 18-45 45-19q26 0 45 19t19 45-19 44.5zm1294 657q20 17.5 20 46.5 0 25-19 44t-46 19q-23 0-43-17.5t-20-45.5q0-27 18-45.5t45-18.5q25 0 45 17.5zM594.5 1046q-20.5 17-43.5 17-28 0-45.5-20t-17.5-43q0-27 19-46t44-19q29 0 46.5 19.5T615 1000q0 29-20.5 46zM912 2225q-20 17-43 17-28 0-46.5-19.5T804 2178q0-28 20-45.5t43-17.5q28 0 46.5 18t18.5 46q0 29-20 46zm-316.5-409.5Q615 1833 615 1863q0 23-17.5 43t-46.5 20q-25 0-44-18.5t-19-46.5 19.5-45.5T551 1798q25 0 44.5 17.5zm1612-340Q2189 1494 2162 1494q-26 0-44.5-18.5T2099 1431t18.5-45 44.5-19q27 0 45.5 19t18.5 45-18.5 44.5zm-863.5.5q19-18 19-45 0-26-19-45t-45-19q-23 0-43 18t-20 46q0 25 17.5 44t45.5 19q26 0 45-18z",
            },
            "children": Array [],
            "name": "path",
          },
        ],
        "viewBox": "0 0 2600 2760.837",
      }
    }
  >
    Published 
    in 49 years
  </StatsItem>
</styled.div>
`;

exports[`AddDependencySearchResult component Rendering should render a header (currentStatus idle) 1`] = `
<styled.header>
  <styled.div>
    <ExternalLink
      color="#304FFE"
      display="inline"
      hoverColor="#3f6cff"
      href="https://www.npmjs.org/package/React"
    >
      <styled.span
        size="small"
      >
        React
      </styled.span>
    </ExternalLink>
    <styled.div
      inline={true}
      size={15}
    />
    <styled.span>
      v
      16.0.0
    </styled.span>
  </styled.div>
  <styled.div>
    <Icon
      fill="currentColor"
      icon={
        Object {
          "attribs": Object {
            "fill": "none",
            "stroke": "currentColor",
            "stroke-linecap": "round",
            "stroke-linejoin": "round",
            "stroke-width": "2",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "points": "20 6 9 17 4 12",
              },
              "children": Array [],
              "name": "polyline",
            },
          ],
          "viewBox": "0 0 24 24",
        }
      }
      size={24}
      style={
        Object {
          "color": "#00E676",
        }
      }
    />
    <styled.div
      size={6}
    />
    Installed
  </styled.div>
</styled.header>
`;

exports[`AddDependencySearchResult component Rendering should render a header (currentStatus installing) 1`] = `
<styled.header>
  <styled.div>
    <ExternalLink
      color="#304FFE"
      display="inline"
      hoverColor="#3f6cff"
      href="https://www.npmjs.org/package/React"
    >
      <styled.span
        size="small"
      >
        React
      </styled.span>
    </ExternalLink>
    <styled.div
      inline={true}
      size={15}
    />
    <styled.span>
      v
      16.0.0
    </styled.span>
  </styled.div>
  <styled.div>
    <Spinner
      size={24}
    />
    <styled.div
      size={6}
    />
    Installingâ€¦
  </styled.div>
</styled.header>
`;

exports[`AddDependencySearchResult component Rendering should render a header (default) 1`] = `
<styled.header>
  <styled.div>
    <ExternalLink
      color="#304FFE"
      display="inline"
      hoverColor="#3f6cff"
      href="https://www.npmjs.org/package/React"
    >
      <styled.span
        size="small"
      >
        React
      </styled.span>
    </ExternalLink>
    <styled.div
      inline={true}
      size={15}
    />
    <styled.span>
      v
      16.0.0
    </styled.span>
  </styled.div>
  <StrokeButton
    fillColor="#FFF"
    onClick={[Function]}
    showStroke={true}
    size="small"
    strokeColors={
      Array [
        "#00C853",
        "#69db0d",
      ]
    }
    textColor="#00C853"
  >
    Add To Project
  </StrokeButton>
</styled.header>
`;

exports[`AddDependencySearchResult component should create a colored download number 4999 1`] = `"#cc004a"`;

exports[`AddDependencySearchResult component should create a colored download number 49999 1`] = `"#FF9100"`;

exports[`AddDependencySearchResult component should create a colored download number 50000 1`] = `"#00C853"`;

exports[`AddDependencySearchResult component should render a StatsItem 1`] = `
<styled.span>
  <Icon
    fill="currentColor"
    icon={
      <Icon
        fill="currentColor"
        icon={
          Object {
            "children": Array [
              Object {
                "attribs": Object {
                  "d": "M2367 2413q21 0 35.5 14.5t14.5 34.5q0 21-14.5 35.5T2367 2512H232q-21 0-35.5-14.5T182 2462V342q0-21 14.5-35.5T232 292q20 0 34.5 14.5T281 342v1644l236-241q19-21 47-21 26 0 46 16l211 177 435-1318q7-18 23-32t36-14q24 0 39.5 9t26.5 26l310 561 179-518q9-21 26-35l392-272q19-13 40-13 26 0 47 19.5t21 49.5q0 17-7 31t-22 25l-374 262-224 641q-6 19-22.5 33t-36.5 14q-28 0-42.5-9.5T1644 1350l-307-558-416 1265q-8 20-25.5 34t-40.5 14q-25 0-44-17l-241-201-289 295v231h2086z",
                },
                "children": Array [],
                "name": "path",
              },
            ],
            "viewBox": "0 0 2600 2760.837",
          }
        }
        size={16}
      />
    }
    size={24}
  />
  <styled.div
    size={6}
  />
  10 downloads a month
</styled.span>
`;
